### ISSUE TITLE
The error propagation rule is Rashid's shortcut, not backpropagation -- and at depth it is measurably better

### ISSUE THREAD

**Issue by xoloki** [xoloki/jlib#141]
`ai::NeuralNetwork` propagates error backward as `W^T e` and applies the
activation derivative only when forming a weight update:

```
  e_out = target - a_out
  e_l   = W_{l+1}^T e_{l+1}                    <- no derivative here
  dW_l  = rate * (e_l * f'(a_l)) a_{l-1}^T     <- only here
```

Backpropagation carries the derivative *with* the error:

```
  d_out = (target - a_out) * f'(z_out)
  d_l   = (W_{l+1}^T d_{l+1}) * f'(z_l)
  dW_l  = rate * d_l a_{l-1}^T
```

So a layer's update here carries **one** derivative factor where the true
gradient carries the product of every factor between that layer and the loss.

## Where it comes from

Rashid, *Make Your Own Neural Network*, which this class was written from:

```python
output_errors = targets - final_outputs
hidden_errors = numpy.dot(self.who.T, output_errors)      # no derivative
self.who += self.lr * numpy.dot((output_errors * final_outputs * (1.0 - final_outputs)), ...)
self.wih += self.lr * numpy.dot((hidden_errors * hidden_outputs * (1.0 - hidden_outputs)), ...)
```

The C++ matches it line for line, so this is a faithful translation and not a
transcription error. **But the book only has one hidden layer**, where the
omission costs a single factor -- `f'(a_out)` missing from the hidden update.
`m_deep` extends the pattern to arbitrary depth, which compounds the omission
once per layer. The book never had to justify that because it never went there.

## Measured

Textbook backpropagation implemented behind the same backend interface and run
against the book's rule. MNIST, 3 epochs, batch 128, best rate per
configuration:

```
                            book's rule   textbook
  sigmoid  1 x 64              94.07%      92.77%
  sigmoid  2 x 64              95.47%      90.63%
  sigmoid  3 x 64              91.93%      25.72%
  relu     3 x 64              95.22%      73.53%
  relu     128 256 128         95.69%      78.56%
  relu     128 256 256 128     89.13%      41.57%
  relu     64 128 256 128 64   84.27%      11.35%
```

**The shortcut wins everywhere, and the gap widens with depth.**

The mechanism is straightforward once stated: correct backpropagation
multiplies the signal by `f'` at every layer on the way down -- at most 0.25
for a sigmoid, and often far less -- while this rule does not. The error
reaching the early layers is therefore much larger here, even though it is not
the gradient of any loss.

## What this means for #131

#131 attributes the depth limit to vanishing gradients through the sigmoid.
That is still the right family of explanation, but the attenuation measured
there (17x across six layers) comes from the `W^T` products and a *single*
derivative, not from a chain of them. Under real backpropagation the same
network attenuates far harder -- which the third and sixth rows above show
directly.

So the network trains at depth **partly because its error rule is wrong**, in a
direction that happens to help.

## Not a request to change it

Recorded so that nobody corrects this on principle and makes the network worse.
If it is ever revisited, the honest framing is that these are two different
algorithms rather than one being a broken version of the other, and the choice
should be made on measurements like the above rather than on which one has a
derivation.

What would be worth trying, and is not this: the modern answers to the problem
the shortcut works around -- residual connections, or normalisation between
layers -- which let correct gradients survive depth rather than avoiding them.

Measured while wiring the network onto ai::backend, which made the comparison
cheap: implementing the alternative was thirty lines against one interface, and
a run takes five seconds on the GPU rather than five minutes on the CPU.
